api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirObservationDaoImpl.java

```diff
@@ -73,6 +111,26 @@ protected String paramToProp(@NotNull String paramName) {
 		}
 	}
 	
+	private void handleValueCodedConcept(Criteria criteria, TokenAndListParam valueConcept) {
+		if (valueConcept != null) {
+			criteria.createAlias("valueCoded", "c");
+			
+			handleAndListParamBySystem(valueConcept, (system, tokens) -> {
```

Comment:
Rather than duplicating code all over the place, we should look for ways to create common methods, e.g. adding something like this to `BaseDaoImpl`:

```
protected Optional<Criterion> findMatchingConcepts(@NotNull String conceptAlias, TokenAndListParam concepts) {
    if (concepts == null) {
         return Optional.empty();
    }

    return handleAndListParamBySystem(concepts, (system, tokens) -> {
	    if (system.isEmpty()) {
			return Optional.of(
			    or(in(String.format("%.conceptId", conceptAlias), tokensToParams(tokens).map(NumberUtils::toInt).collect(Collectors.toList())),
			        in(String.format("%.uuid", conceptAlias), tokensToList(tokens))));
		} else {
			if (!containsAlias(criteria, "cm")) {
				criteria.createAlias(String.format("%.conceptMappings", conceptAlias), "cm").createAlias("cm.conceptReferenceTerm", "crt");
			}

			return Optional.of(generateSystemQuery(system, tokensToList(tokens)));
		}
	}
}
```

And then just changing this to:

```java
private void handleValueCodedConcept(Criteria criteria, TokenAndListParam valueConcepts) {
	if (valueConcept != null) {
		criteria.createAlias("valueCoded", "c");
        findMatchingConcepts("c", valueConcepts).ifPresent(criteria::add);
    }
}
```

Do similar things for `FhirObservationDaoImpl` and `FhirAllergyIntoleranceDaoImpl`.

This would also address my comment above about refactoring.